ddtrace/tracer.py

```diff
@@ -518,26 +536,45 @@ def start_span(
             ):
                 span._set_str_tag(VERSION_KEY, config.version)
 
-        # add it to the current context
-        context.add_span(span)
-
         # update set of services handled by tracer
         if service and service not in self._services and self._is_span_internal(span):
             self._services.add(service)
 
+        if activate:
+            self.context_provider.activate(span)
+
         for p in self._span_processors:
             p.on_span_start(span)
 
         self._hooks.emit(self.__class__.start_span, span)
         return span
 
     def _on_span_finish(self, span):
-        if not self.enabled:
```

Comment:
I don't know if we have such tests but it might be easy to test for memory leaks by testing this special test cases and seeing what's left in the tracer/writer once you're done and "flush" things out.

Reply:
:+1: agreed - it'd be really valuable to have these (we don't).

Reply:
I think I'm going to make having these tests be mandatory before merging this PR. Especially because of e2456e9

Reply:
This was done in #2497 and #2493